b2g/components/RemoteControlService.jsm

```diff
@@ -0,0 +1,395 @@
+    DEBUG && debug("doStart");
+
+    if (this._socket) {
+      aReject("SocketAlreadyInit");
+      return;
+    }
+
+    this._doQuit = this._socketClosed = false;
+
+    // Monitor xpcom-shutdown to stop service and clean up
+    Services.obs.addObserver(this, "xpcom-shutdown", false);
+
+    // Start TLSSocketServer with self-signed certification
+    Cc["@mozilla.org/psm;1"].getService(Ci.nsISupports);
+    certService.getOrCreateCert("RemoteControlService", {
+      handleCert: function(cert, result) {
+        if(result) {
+          aReject("getOrCreateCert " + result);
+        } else {
+          let self = RemoteControlService;
+
+          try {
+            // Try to get random port
+            let ios = Cc["@mozilla.org/network/io-service;1"].getService(Ci.nsIIOService);
+            let socket;
+            for (let i = 100; i; i--) {
+              let temp = Cc["@mozilla.org/network/tls-server-socket;1"].createInstance(Ci.nsITLSServerSocket);
+              temp.init(self._port, false, MAX_CLIENT_CONNECTIONS);
+              temp.serverCert = cert;
+
+              let allowed = ios.allowPort(temp.port, "tls");
+              if (!allowed) {
+                DEBUG && debug("Warning: obtained TLSServerSocket listens on a blocked port: " + temp.port);
+              }
+
+              if (!allowed && self._port == -1) {
+                DEBUG && debug("Throw away TLSServerSocket with bad port.");
+                temp.close();
+                continue;
+              }
+
+              socket = temp;
+              break;
+            }
+
+            if (!socket) {
+              throw new Error("No socket server available. Are there no available ports?");
+            }
+
+            DEBUG && debug("Listen on port " + socket.port + ", " + MAX_CLIENT_CONNECTIONS + " pending connections");
+
+            socket.serverCert = cert;
+            socket.setSessionCache(false);
+            socket.setSessionTickets(false);
+            socket.setRequestClientCertificate(Ci.nsITLSServerSocket.REQUEST_NEVER);
+
+            socket.asyncListen(self);
+            self._port = socket.port;
+            self._socket = socket;
+          } catch (e) {
+            DEBUG && debug("Could not start server on port " + self._port + ": " + e);
+            aReject("Start TLSSocketServer fail");
+          }
+
+          // Register mDNS remote control service with this._port
+          if (("@mozilla.org/toolkit/components/mdnsresponder/dns-sd;1" in Cc)) {
+            let serviceInfo = Cc["@mozilla.org/toolkit/components/mdnsresponder/dns-info;1"].createInstance(Ci.nsIDNSServiceInfo);
+            serviceInfo.serviceType = "_remotecontrol._tcp";
+            serviceInfo.serviceName = Services.prefs.getCharPref("dom.presentation.device.name");
+            serviceInfo.port = self._port;
+
+            let mdns = Cc["@mozilla.org/toolkit/components/mdnsresponder/dns-sd;1"].getService(Ci.nsIDNSServiceDiscovery);
```

Comment:
nit: line break before `.getService`.

Or you might consider make these contractId strings global const